### DISCUSSION TITLE
Makie as interactive 3D display for neutrino events... how to use shaders?

### DISCUSSION THREAD

**Discussion by tamasgal** [JuliaPlots/Makie.jl#3030] [Q&A]
Hi all,

I wrote an interactive 3D display (PyOpenGL and GLUT) a long time ago and now I want to revive the project in pure Julia. This is how the old project looks like:

<img width="993" alt="Screenshot 2023-06-27 at 13 41 51" src="https://github.com/MakieOrg/Makie.jl/assets/1730350/e9e3c45c-f169-4f1f-821b-da355cac0078">

I already played around with GLMakie and managed to draw some basic 3D scenes and rotate the scene etc. but I am struggling with a few things.

The basic idea behind such an event display is very easy: display the detector geometry, which is a few thousand optical modules in a 3D grid, show some lines as particles travel and use some graphics (spheres, flashes) to indicate hits on the photomultiplier tubes as time passes. The user should be able to rotate, move and zoom the scene.

Anyways, long story short, here are some basic questions:

- is (GL)Makie.jl suitable for rendering thousands of objects or should I better start from scratch with a stripped slim self-made OpenGL framework?
- I could not figure out how to use custom shaders to e.g. display the detector modules. In the screenshot above, those are only dots but I'd like to use custom shaders to make them look much more realistic
- same for indicating signals on the photomultiplier tubes, instead of spheres made out of polygons (which are computationally intensive) I'd like to use shaders

Sorry if the questions are so vague, I just wanted to have some feedback from Makie experts since I am a bit stuck ;)

If there was some easy-to-understand example to e.g. display a set of objects in 3D, which are displayed via a custom shader, that would be awesome. Of course once I figured it out, I am happy to provide PR(s) to the documentation!

**Comment by SimonDanisch** [comment]
The existing shaders are pretty customizable (scatter/meshscatter), so I think there's a chance that you can visualize the particles without needing a custom shaders.
Thousands of objects aren't a problem, you should be able to visualize at least millions of particles interactively with GLMakie.
If you really need custom shaders, there is a way to do that, but I'd need to write some small tutorial I suppose, since it isn't documented and not that easy to figure out yourself.

**Reply by tamasgal** [reply]
Thanks Simon, I'll play around with the existing shaders.

For visualising the optical modules, I need a more complex object, but I'll see how far I get. I'd be very happy to read a tutorial about custom shaders, or even a simple example which shows how to pass a vertex shader and a fragment shader to render with a given camera...?

Glad to hear that millions of particles are doable ;)
